src/Tolerance/Operation/Runner/RetryPromiseOperationRunner.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+/*
+ * This file is part of the Tolerance package.
+ *
+ * (c) Samuel ROZE <samuel.roze@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Tolerance\Operation\Runner;
+
+use Tolerance\Operation\Exception\UnsupportedOperation;
+use Tolerance\Operation\Operation;
+use Tolerance\Operation\PromiseOperation;
+use Tolerance\Operation\RetryEvaluator\AlwaysRetryEvaluator;
+use Tolerance\Operation\RetryEvaluator\NeverRetryEvaluator;
+use Tolerance\Operation\RetryEvaluator\RetryEvaluator;
+use Tolerance\Waiter\StatefulWaiter;
+use Tolerance\Waiter\Waiter;
+use Tolerance\Waiter\WaiterException;
+
+class RetryPromiseOperationRunner implements OperationRunner
+{
+    /**
+     * @var Waiter
+     */
+    private $waitStrategy;
+    /**
+     * @var RetryEvaluator
+     */
+    private $fulfilledEvaluator;
+    /**
+     * @var RetryEvaluator
+     */
+    private $rejectedEvaluator;
+
+    /**
+     * @param Waiter $waitStrategy
+     * @param RetryEvaluator|null $fulfilledEvaluator
+     * @param RetryEvaluator|null $rejectedEvaluator
+     */
+    public function __construct(Waiter $waitStrategy, RetryEvaluator $fulfilledEvaluator = null, RetryEvaluator $rejectedEvaluator = null)
+    {
+        $this->waitStrategy = $waitStrategy;
+        $this->fulfilledEvaluator = $fulfilledEvaluator ?: new NeverRetryEvaluator();
```

Comment:
Guzzle actually reject the promise for 4xx & 5xx responses. This is done by the [`httpErrors` middleware](https://github.com/guzzle/guzzle/blob/122c0b7f784de7be0d4098a093d9c3debc72b926/src/Middleware.php#L52-L70) that is [by default in the `HandlerStack`](https://github.com/guzzle/guzzle/blob/122c0b7f784de7be0d4098a093d9c3debc72b926/src/HandlerStack.php#L41).